### ISSUE TITLE
Bug with checkpointing and mapped tasks

### ISSUE THREAD

**Issue by emcake** [PrefectHQ/prefect#3219]
## Description

Currently restarting a failed flow that has failed on some mapped tasks seems wrong in the following ways:

1. Previously successful mapped childs will re-run
2. `run_count` doesn't seem to change between restarts

## Expected Behavior

1. A previously successful mapped task should not be rerun by checkpointing if it has already succeeded
2. the `run_count` should up between runs.

## Reproduction

Here is a minimal flow that can reproduce this is below. Start a cloud run, it will fail on the first mapped task, but the other two will succeed. Use the 'restart' button in the top right of the UI to restart this flow-run. Note that previously successful mapped runs run again. Note that the first instance fails again, and (in the logs) `run_count` is still `1`, even though the UI clearly shows multiple runs.

```python

import prefect

@prefect.task
def task(x):
    prefect.context.get('logger').info(f'x = {x} and run_count = {prefect.context.get("task_run_count")}')
    if x == 0 and prefect.context.get('task_run_count') == 1:
        raise prefect.engine.signals.FAIL('fail on first attempt')
    else:
        return x + 1

@prefect.task
def summy(xs):
    return sum(xs)

with prefect.Flow('test-restarts') as flow:
    list = prefect.tasks.core.constants.Constant([0,1,2])

    mapped = task.map(list)

    summy(mapped)

flow.register(project_name='mattk')
```

## Environment

```
root@docker-desktop:/workspaces/prefect# prefect diagnostics
{
  "config_overrides": {},
  "env_vars": [],
  "system_information": {
    "platform": "Linux-4.19.104-microsoft-standard-x86_64-with-debian-10.5",
    "prefect_version": "0.13.2",
    "python_version": "3.7.9"
  }
}
```

**Comment by cicdw** [comment]
Hi @emcake - `task_run_count` in context is a tracker for automatic retries, so from a Prefect-mechanics point of view it makes sense that this was not incremented.  Cloud actually stores the "real" value that you are looking for as a queryable field on the `task_run` route.  E.g.,
```graphql
query{
  flow_run_by_pk(id: "ZZZ") {
    task_runs{
      run_count # will return 2 for some of the runs in this flow
    }
  }
}
```
I recognize that the naming could be a little confusing here, so that's something we might consider updating.

However, the fact that your mapped tasks re-ran is definitely a bug - I would have only expected map_index =0 and the `summy` task to rerun (and still fail).

**Comment by emcake** [comment]
yeah, I did wonder if it was something like that :) That was just my attempt to make it succeed on the retry, so ignore that part.
